### ISSUE TITLE
Question: sampling from trained encoder-decoder-coupling model

### ISSUE THREAD

**Issue by rahul-iisc** [Element-Research/rnn#247]
Hi,

I am trying to improve on the encoder-decoder-coupling example while learning. I generated some synthetic data for training. The task is to repeat each element of source sequence twice. e.g. {1,2,3,4} should translate to {1,1,2,2,3,3,4,4}. The training was fine and I was able to get the validation loss from initial 69 to around 3.
Next, I tried to sample using the saved model but sampling just gave me first character of the input sequence repeatedly.

the training code is similar to the example code. Below is the code for sampling (taken from #142) -
```lua
function sample(input, model, maxOut)
   enc:evaluate()
   dec:evaluate()

   encLSTM:remember()
   decLSTM:remember()

    enc:forward(input)
    forwardConnect(encLSTM, decLSTM)

    local sequence = {}

    -- -- Forward <go> and all of it's output recursively back to the decoder
    local decInSeq = torch.Tensor{GO}    -- first element in 'decInSeq'
    for counter = 1, maxOut do           -- sample maximum maxOut elements
        local prediction = model.dec:forward(decInSeq )[1]

        -- prediction contains the probabilities for each element IDs.
        -- The index of the probability is the word ID.
        local prob, preds = prediction:sort(1, true)

        -- The first one is the most likely.
        decInSeq[1] = preds[1]                 -- update decInSeq with current prediction
        print(preds[1], preds[2], preds[3])    -- they all give same element in each step

        -- Terminate on EOS token
        if decInSeq[1] == EOS then break end

        table.insert(sequence, preds[1])
    end 

    model.encoder:forget()
    model.decoder:forget()

    return sequence
end
```

Could anyone please comment if the sampling code looks right? or if something else could go wrong. Thanks in advance.

**Comment by nicholas-leonard** [comment]
I think you forgot to set `output[1] = preds[1]` inside your for loop (at the end). Otherwise, you are repeatedly presenting the same input multiple times.

I would also use `torch.multinomial` to sample predictions instead of `torch.sort`.

**Comment by rahul-iisc** [comment]
@nicholas-leonard I intended to change output to decInSeq but forgot to rename it at a few places, sorry about that. As I'm updating decInSeq and in for loop's first line I forward decInSeq in decoder, so it won't be the same input each time.

**Comment by rahul-iisc** [comment]
@nicholas-leonard I checked by passing the full decIn sequence and then the prediction comes out correct, it seems in order to sample without providing decInSeq, I'll have to train the network without teacher forcing.

**Comment by rahul-iisc** [comment]
@nicholas-leonard 
I am working on this seq2seq example and everything I learned till now I've added on this repo: https://github.com/rahul-iisc/seq2seq-mapping. I'm using variable length sequences with optimization using optim package and some more small add-ons. If possible can you give a link to this in readme for encoder-decoder-coupling example.

**Comment by nicholas-leonard** [comment]
@rahul-iisc Nice. I added the link to your repo : https://github.com/Element-Research/rnn/commit/18c4efce04c2b0a297be011c8057e92d4ad53391. Thanks, keep at it!
